native/src/gpxRouteApproximation.cpp

```diff
@@ -0,0 +1,524 @@
+						 SHARED_PTR<GpxPoint>& start, SHARED_PTR<GpxPoint>& target, bool prevRouteCalculated) {
+	vector<SHARED_PTR<RouteSegmentResult>> res;
+	bool routeIsCorrect = false;
+	if (start->pnt && target->pnt) {
+		start->pnt = std::make_shared<RouteSegmentPoint>(start->pnt);
+		target->pnt = std::make_shared<RouteSegmentPoint>(target->pnt);
+		// OsmAnd::LogPrintf(OsmAnd::LogSeverityLevel::Info, "[Native] Calculate: %.5f %.5f -> %.5f %.5f",
+		// 				  start->pnt->getPreciseLatLon().lat, start->pnt->getPreciseLatLon().lon,
+		// 				  target->pnt->getPreciseLatLon().lat, target->pnt->getPreciseLatLon().lon);
+		gctx->routeDistCalculations += (target->cumDist - start->cumDist);
+		gctx->routeCalculations++;
+		RoutingContext* cp = new RoutingContext(gctx->ctx);
+		res = router->searchRouteInternalPrepare(cp, start->pnt, target->pnt, nullptr);
+		delete cp;
+		// BinaryRoutePlanner.printDebugMemoryInformation(gctx.ctx);
+		routeIsCorrect = !res.empty();
+		for (int k = start->ind + 1; routeIsCorrect && k < target->ind; k++) {
+			SHARED_PTR<GpxPoint>& ipoint = gpxPoints[k];
+			if (!pointCloseEnough(gctx, ipoint, res)) {
+				routeIsCorrect = false;
+			}
+		}
+		if (routeIsCorrect) {
+			// correct start point though don't change end point
+			if (prevRouteCalculated) {
+				if (res[0]->object->getId() == start->pnt->getRoad()->getId()) {
+					// start point could shift to +-1 due to direction
+					res[0]->setStartPointIndex(start->pnt->getSegmentEnd());
+					if (res[0]->object->getPointsLength() != start->pnt->getRoad()->getPointsLength()) {
+						res[0]->object = start->pnt->getRoad();
+					}
+					if (res[0]->getStartPointIndex() == res[0]->getEndPointIndex()) {
+						res.erase(res.begin());
+					}
+				} else {
+					// for native routing this is possible when point lies on intersection of 2 lines
+					// solution here could be to pass to native routing id of the route
+					// though it should not create any issue
+					// System.out.println("??? not found " + start.pnt.getRoad().getId() + " instead "
+					// + res.get(0).getObject().getId());
+				}
+			}
+			for (SHARED_PTR<RouteSegmentResult>& seg : res) {
+				seg->setGpxPointIndex(start->ind);
+			}
+			start->routeToTarget = res;
+			start->targetInd = target->ind;
+			// OsmAnd::LogPrintf(OsmAnd::LogSeverityLevel::Info, "[Native] Calculate res: %.5f %.5f -> %.5f %.5f",
+			// 				  res[0]->getStartPoint().lat, res[0]->getStartPoint().lon,
+			// 				  res[res.size() - 1]->getEndPoint().lat, res[res.size() - 1]->getEndPoint().lon);
+		}
+	}
+	return routeIsCorrect;
+}
+
+
+double GpxRouteApproximation::distFromLastPoint(double lat, double lon) {
+	if (fullRoute.size() > 0) {
+		return getDistance(getLastPoint().lat, getLastPoint().lon, lat, lon);
+	}
+	return 0;
+}
+
+LatLon GpxRouteApproximation::getLastPoint() {
+	return fullRoute[fullRoute.size() - 1]->getEndPoint();
```

Comment:
Possible crash

Reply:
There is no better way in this particular case. C++ version of getLastPoint must return LatLon and can not return null (when it is possible in Java).

To avoid a potential crash, both callers of the function are protected with gctx->fullRoute.size() > 0 check.

This code was double-checked few months before.

Reply:
Added exception throw